pvlib/spectrum/mismatch.py

```diff
@@ -235,3 +236,158 @@ def integrate(e):
+
+    where ``c``, ``a`` and ``b`` are the model parameters, different for each
+    irradiance component.
+
+    References
+    ----------
+    .. [1] Martín, N. and Ruiz, J.M. (1999), A new method for the spectral
+       characterisation of PV modules. Prog. Photovolt: Res. Appl., 7: 299-310.
+       :doi:10.1002/(SICI)1099-159X(199907/08)7:4<299::AID-PIP260>3.0.CO;2-0
+
+    See Also
+    --------
+    pvlib.irradiance.clearness_index
+    pvlib.atmosphere.get_relative_airmass
+    pvlib.atmosphere.get_absolute_airmass
+    pvlib.atmosphere.first_solar_spectral_correction
+    """
+    # Note tests for this function are prefixed with test_martin_ruiz_mm_*
+
+    IRRAD_COMPONENTS = ('poa_direct', 'poa_sky_diffuse', 'poa_ground_diffuse')
+    # Fitting parameters directly from [1]_
+    MARTIN_RUIZ_PARAMS = pd.DataFrame(
+        index=('monosi', 'polysi', 'asi'),
+        columns=pd.MultiIndex.from_product([IRRAD_COMPONENTS,
+                                           ('c', 'a', 'b')]),
+        data=[  # Direct(c,a,b)  | Sky diffuse(c,a,b) | Ground diffuse(c,a,b)
+            [1.029, -.313, 524e-5, .764, -.882, -.0204, .970, -.244, .0129],
+            [1.029, -.311, 626e-5, .764, -.929, -.0192, .970, -.270, .0158],
+            [1.024, -.222, 920e-5, .840, -.728, -.0183, .989, -.219, .0179],
+        ])
+
+    # Argument validation and choose components and model parameters
+    if cell_type is not None and model_parameters is None:
+        # Infer parameters from cell material
+        cell_type_lower = cell_type.lower()
+        if cell_type_lower in MARTIN_RUIZ_PARAMS.index:
+            _params = MARTIN_RUIZ_PARAMS.loc[cell_type_lower]
+        else:
+            raise NotImplementedError('Cell type parameters not defined in '
+                                      'algorithm. Allowed types are '
+                                      f'{tuple(MARTIN_RUIZ_PARAMS.index)}')
+    elif cell_type is None and model_parameters is None:
+        raise TypeError('You must pass at least "cell_type" '
+                        'or "model_parameters" as arguments.')
+    elif model_parameters is not None:  # Use user-defined model parameters
+        # Validate 'model_parameters' sub-dicts keys
+        if any([{'a', 'b', 'c'} != set(model_parameters[component].keys())
+                for component in model_parameters.keys()]):
+            raise ValueError("You must specify model parameters with keys "
+                             "'a','b','c' for each irradiation component.")
+
+        _params = model_parameters
+        if cell_type is not None:
+            warn('Both "cell_type" and "model_parameters" given. '
+                 'Using provided "model_parameters".')
+
+    # Compute difference here to avoid recalculating inside loop
+    kt_delta = clearness_index - 0.74
+    am_delta = airmass_absolute - 1.5
+
+    if hasattr(kt_delta, '__iter__') or hasattr(am_delta, '__iter__'):
```

Comment:
Checking for the presence of `.__iter__` is not something we do elsewhere in pvlib (to my knowledge); in cases where we can't write code that automatically returns the right types without intervention, we usually check `np.isscalar` or the types of the inputs.  Can we do the same here?